File: Editorial/2001-2100/2083_Substrings_That_Begin_and_End_With_the_Same_Letter/Java/1_Prefix_Count.java
class Solution {

    public long numberOfSubstrings(String s) {
        int n = s.length();
        // Number of substrings can exceed the integer range
        long answer = 0;
        long[] prefixCount = new long[26];

        for (int i = 0; i < s.length(); i++) {
            // Increment the number of times we encountered the current letter
            // so far.
            prefixCount[s.charAt(i) - 'a']++;

            // Current letter can be paired with all the occurrences of it that
            // comes before, including itself, to form a valid substring.
            answer += prefixCount[s.charAt(i) - 'a'];
        }

        return answer;
    }
}

File: Editorial/2001-2100/2083_Substrings_That_Begin_and_End_With_the_Same_Letter/CPP/1_Prefix_Count.cpp
class Solution {
public:
    long long numberOfSubstrings(string s) {
        int n = s.size();
        // Number of substrings can exceed the integer range
        long long answer = 0;
        vector<long long> prefixCount(26, 0);

        for (int i = 0; i < s.size(); i++) {
            // Increment the number of times we encountered the current letter
            // so far.
            prefixCount[s[i] - 'a']++;

            // Current letter can be paired with all the occurrences of it that
            // comes before, including itself, to form a valid substring.
            answer += prefixCount[s[i] - 'a'];
        }

        return answer;
    }
};
